ops/ruby/devops-framework/tasks/lib/service_account_manager.rb

```diff
@@ -0,0 +1,81 @@
+require "json"
+require 'tmpdir'
+require 'fileutils'
+require 'logger'
+
+# Entering a service account context ensures that a keyfile exists at the given
+# path for the given service account, and that GOOGLE_APPLICATION_CREDENTIALS is
+# pointing to it (for application default credentials). Creates this SA key and
+# file if necessary, and destroys it when leaving the context.
+#
+# Nested service account contexts are not supported.
+class ServiceAccountManager
+
+  SERVICE_ACCOUNT_KEY_FILE_NAME = "sa-key.json"
+
+  def initialize(project, service_account, logger = Logger.new(STDOUT))
+    @logger = logger
+    @project = project
+    @service_account = service_account
+  end
+
+  attr_reader :project
+  attr_reader :service_account
+  attr_reader :credentials_path
+
+  def run()
+    @logger.info("service_account = #{@service_account}")
+    credentials_path = create_credentials_file
+
+    @logger.info("Setting environment variable GOOGLE_APPLICATION_CREDENTIALS to #{credentials_path}")
+    ENV["GOOGLE_APPLICATION_CREDENTIALS"] = credentials_path
+
+    begin
+      yield self
```

Comment:
what an interesting pattern

Reply:
Yeah, this was the first language I tried to use professionally in like 2007, and it was the wrong choice for me at the time. I couldn't wrap my head around lambdas until years later, and Ruby has half a dozen ways of doing things like this.